packages/rocketchat-irc/server/server.js

```diff
@@ -0,0 +1,455 @@
+
+		this.isJoiningRoom = false;
+		roomName = this.pendingJoinRoomBuf.shift();
+		if (roomName) {
+			this.joinRoom({
+				t: 'c',
+				name: roomName
+			});
+		}
+	}
+
+	sendRawMessage(msg) {
+		console.log('[irc] sendRawMessage -> '.yellow, msg.slice(0, -2));
+		if (this.isConnected) {
+			this.socket.write(msg);
+		} else {
+			this.msgBuf.push(msg);
+		}
+	}
+
+	sendMessage(room, message) {
+		console.log('[irc] sendMessage -> '.yellow, 'userName:', message.u.username);
+		let target = '';
+		if (room.t === 'c') {
+			target = `#${ room.name }`;
+		} else if (room.t === 'd') {
+			const usernames = room.usernames;
+			usernames.forEach(name => {
+				if (message.u.username !== name) {
+					target = name;
+					return;
+				}
+			});
+		}
+		const cacheKey = [this.user.username, target, message.msg].join(',');
+		console.log('[irc] ircSendMessageCache.set -> '.yellow, 'key:', cacheKey, 'ts:', message.ts.getTime());
+		ircSendMessageCache.set(cacheKey, message.ts.getTime());
+		const msg = `PRIVMSG ${ target } :${ message.msg }\r\n`;
+		this.sendRawMessage(msg);
+	}
+
+	initRoomList() {
+		const roomsCursor = RocketChat.models.Rooms.findByTypeContainingUsername('c', this.user.username, { fields: { name: 1, t: 1 }});
+		const rooms = roomsCursor.fetch();
+
+		rooms.forEach(room => {
+			this.joinRoom(room);
+		});
+	}
+
+	joinRoom(room) {
+		let msg;
+		if (room.t !== 'c' || room.name === 'general') {
+			return;
+		}
+		if (this.isJoiningRoom) {
+			this.pendingJoinRoomBuf.push(room.name);
+		} else {
+			console.log('[irc] joinRoom -> '.yellow, 'roomName:', room.name, 'pendingJoinRoomBuf:', this.pendingJoinRoomBuf.join(','));
+			msg = `JOIN #${ room.name }\r\n`;
+			this.receiveMemberListBuf[room.name] = [];
+			this.sendRawMessage(msg);
+			this.isJoiningRoom = true;
+		}
+	}
+
+	leaveRoom(room) {
+		if (room.t !== 'c') {
+			return;
+		}
+		const msg = `PART #${ room.name }\r\n`;
+		this.sendRawMessage(msg);
+	}
+
+	getMemberList(room) {
+		if (room.t !== 'c') {
+			return;
+		}
+		const msg = `NAMES #${ room.name }\r\n`;
+		this.receiveMemberListBuf[room.name] = [];
+		this.sendRawMessage(msg);
+	}
+
+	onAddMemberToRoom(member, roomName) {
+		if (this.user.username === member) {
+			return;
+		}
+		console.log('[irc] onAddMemberToRoom -> '.yellow, 'roomName:', roomName, 'member:', member);
+		this.createUserWhenNotExist(member);
+		RocketChat.models.Rooms.addUsernameByName(roomName, member);
+	}
+
+	onRemoveMemberFromRoom(member, roomName) {
+		console.log('[irc] onRemoveMemberFromRoom -> '.yellow, 'roomName:', roomName, 'member:', member);
+		RocketChat.models.Rooms.removeUsernameByName(roomName, member);
+	}
+
+	onQuitMember(member) {
+		console.log('[irc] onQuitMember ->'.yellow, 'username:', member);
+		RocketChat.models.Rooms.removeUsernameFromAll(member);
+		Meteor.users.update({ name: member }, { $set: { status: 'offline' }});
+	}
+
+	createUserWhenNotExist(name) {
+		let user;
```

Comment:
this logic is seem good, but at least use one line to declare and set the variable.